### PR TITLE
feat(manager/sbt): support replacement

### LINKED ISSUE #25643: Support autoreplace for sbt manager
### Discussed in https://github.com/renovatebot/renovate/discussions/25637

<div type='discussions-op-text'>

<sup>Originally posted by **gaeljw** November  7, 2023</sup>
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Given the following repository: https://github.com/gaeljw/renovate-replacement-play, I'd like to get a PR opened using the "replacement" feature because one of my dependencies got its name changed.

- `"com.typesafe.play" % "sbt-plugin"` is renamed to `"org.playframework" % "sbt-plugin"`

### What you tried so far.

I've added the following replacement rules to my `renovate.json` config:

```json
"packageRules": [
      {
        "matchPackageNames": [
          "com.typesafe.play:sbt-plugin"
        ],
        "replacementName": "org.playframework:sbt-plugin",
        "replacementVersion": "3.0.0"
      }
    ]
```

But so far, I don't get any replacement PR opened. I only get new versions on the "old" naming, I'd like to get additional PRs with the new name and version 3.0.0.

### Relevant debug logs

<details><summary>Logs</summary>

```
"sbt": [
      {
        "deps": [
          {
            "datasource": "sbt-plugin",
            "depName": "com.typesafe.play:sbt-plugin",
            "packageName": "com.typesafe.play:sbt-plugin",
            "currentValue": "2.9.0",
            "depType": "plugin",
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases"
            ],
            "updates": [
              {
                "updateType": "replacement",
                "newName": "org.playframework:sbt-plugin",
                "newValue": "3.0.0",
                "branchName": "renovate/com.typesafe.play-sbt-plugin-replacement"
              }
            ],
            "versioning": "ivy",
            "warnings": [],
            "sourceUrl": "https://github.com/playframework/playframework",
            "registryUrl": "https://repo.maven.apache.org/maven2",
            "dependencyUrl": "https://repo.maven.apache.org/maven2/com/typesafe/play/sbt-plugin",
            "currentVersion": "2.9.0",
            "fixedVersion": "2.9.0"
          }
        ],
        "packageFile": "project/plugins.sbt"
      }
    ]
  }

...

INFO: Cannot find replaceString in current file content. Was it already updated? (branch="renovate/com.typesafe.play-sbt-plugin-replacement")
{
  "packageFile": "project/plugins.sbt"
  "depName": "com.typesafe.play:sbt-plugin"
  "existingContent": "\naddSbtPlugin(\"com.typesafe.play\" % \"sbt-plugin\" % \"2.9.0\")\n"
  "replaceString": "2.9.0"
}

DEBUG: No content changed (branch="renovate/com.typesafe.play-sbt-plugin-replacement")
{
  "packageFile": "project/plugins.sbt"
  "depName": "com.typesafe.play:sbt-plugin"
}
```

</details>

I guess the error `Cannot find replaceString in current file content` is relevant but I'm not sure how to fix it.</div>

### PR DESCRIPTION
**Pull request by gaeljw**
## Changes

Attempts to solve https://github.com/renovatebot/renovate/issues/25643, that is support the replacement feature in sbt.

In order to support replacement, I've had to get rid of the default "auto replace" that was used so far for SBT. Hence this PR contains code to support replacement as well as "regular" version updates.

I've written one test case for each supported way of extracting a dependency according to `extract.ts`. Hence I believe most, if not all, expected usage of SBT currently are covered.

Replaces the previous attempt at https://github.com/renovatebot/renovate/pull/39023.

## Context

Please select one of the below:

- [x] This closes an existing Issue, Closes: #25643
- [ ] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

Did you use AI tools to create any part of this pull request?

Please select one option and, if yes, briefly describe how AI was used (e.g., code, tests, docs) and which tool(s) you used.

- [ ] No — I did not use AI for this contribution.
- [x] Yes — minimal assistance (e.g., IDE autocomplete, small code completions, grammar fixes).
- [ ] Yes — substantive assistance (AI generated non‑trivial portions of code, tests, or documentation).
- [ ] Yes — other (please describe):

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The public repository: https://github.com/gaeljw/renovate-sbt-replacement-3

### CODE REVIEW

**Comment by gaeljw** [comment]
Definitely not high priority but any chance I could get some feedback on this? I'd really like to have support for replacements in SBT.

**Comment by gaeljw** [comment]
Hello folks, bumping this one :)
I'm happy to resolve conflicts that appeared since then but I'll spend some time only if there's a chance this gets merged at some point.

File: lib/modules/manager/sbt/update.ts [hunk 1]
```diff
@@ -13,6 +16,7 @@ export function bumpPackageVersion(
     { bumpVersion, currentValue },
     'Checking if we should bump build.sbt version',
   );
+  logger.warn({ content }, 'bumpPackageVersion content');
```
Comment (RahulGautamSingh) [inline lib/modules/manager/sbt/update.ts]: warn?

File: lib/modules/manager/sbt/update.ts [hunk 2]
```diff
@@ -32,3 +36,75 @@ export function bumpPackageVersion(
...
+
+  // Update the version
+  if (newValue && currentValue && newValue !== currentValue) {
+    if (sharedVariableName) {
+      logger.debug(
+        { depName, sharedVariableName, currentValue, newValue },
+        'Updating version in shared variable for sbt dependency',
+      );
+
+      const patternSharedVariable = new RegExp(
```
Comment (RahulGautamSingh) [inline lib/modules/manager/sbt/update.ts]: use regEx

File: lib/modules/manager/sbt/update.ts [hunk 3]
```diff
@@ -32,3 +36,75 @@ export function bumpPackageVersion(
...
+        `${escapeRegExp(sharedVariableName)}(\\s*:\\s*String)?\\s*=\\s*"${currentValue}"`,
```
Comment (RahulGautamSingh) [inline lib/modules/manager/sbt/update.ts:93]: I don't think we use this. @viceice what is the alternative for this?

**Review by RahulGautamSingh — CHANGES_REQUESTED** [review]

File: lib/modules/manager/sbt/update.ts (hunk 1 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/update.ts]: A leftover of some testing I believe. Removed the log entirely.

File: lib/modules/manager/sbt/update.ts (hunk 3 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/update.ts:93]: Do you mean `escapeRegExp` @RahulGautamSingh ? I've seen it in a few files in the repo.

Comment (viceice) [reply lib/modules/manager/sbt/update.ts:93]: @RahulGautamSingh what do you mean?

Comment (viceice) [reply lib/modules/manager/sbt/update.ts:93]: @RahulGautamSingh what do you mean?

File: lib/modules/manager/sbt/update.ts [hunk 4]
```diff
@@ -32,3 +35,75 @@ export function bumpPackageVersion(
...
+          return `${sharedVariableName}${typed ?? ''} = "${newValue}"`;
+        },
+      );
+    } else {
+      logger.debug(
+        { depName, currentValue, newValue },
+        'Updating version for sbt dependency',
+      );
+
+      const pattern = new RegExp(
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]:
```suggestion
      const pattern = regEx(
```
use our regex helper

File: lib/modules/manager/sbt/update.ts [hunk 5]
```diff
@@ -32,3 +35,75 @@ export function bumpPackageVersion(
...
+
+      const patternSharedVariable = regEx(
+        `${escapeRegExp(sharedVariableName)}(\\s*:\\s*String)?\\s*=\\s*"${currentValue}"`,
+        'g',
+      );
+
+      updatedContent = updatedContent.replace(
+        patternSharedVariable,
+        (match, typed) => {
+          return `${sharedVariableName}${typed ?? ''} = "${newValue}"`;
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: why not doing another replace on match? then you even don't need to care about whitespace and other things

File: lib/modules/manager/sbt/update.ts [hunk 6]
```diff
@@ -32,3 +35,75 @@ export function bumpPackageVersion(
...
+
+      const pattern = new RegExp(
+        `"${escapeRegExp(groupId)}"\\s*(%|%%|%%%)\\s*"${escapeRegExp(artifactId)}"\\s*%\\s*"${escapeRegExp(currentValue)}"`,
+        'g',
+      );
+
+      updatedContent = updatedContent.replace(
+        pattern,
+        (match, percentSigns) => {
+          return `"${groupId}" ${percentSigns} "${artifactId}" % "${newValue}"`;
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: same here

File: lib/modules/manager/sbt/update.ts [hunk 7]
```diff
@@ -32,3 +35,75 @@ export function bumpPackageVersion(
...
+  // Update the dependency/artifact groupId/artifactId in case of replacement
+  if (upgrade.updateType === 'replacement' && newName) {
+    logger.debug(
+      { depName, newName, currentValue, newValue },
+      'Applying replacement update for sbt dependency',
+    );
+
+    const [newGroupId, newArtifactId] = newName.split(':', 2);
+
+    const pattern = new RegExp(
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]:
```suggestion
    const pattern = regEx(
```

File: lib/modules/manager/sbt/update.ts [hunk 8]
```diff
@@ -32,3 +35,75 @@ export function bumpPackageVersion(
...
+      `"${escapeRegExp(groupId)}"\\s*(%|%%|%%%)\\s*"${escapeRegExp(artifactId)}"`,
+      'g',
+    );
+
+    updatedContent = updatedContent.replace(pattern, (match, percentSigns) => {
+      return `"${newGroupId}" ${percentSigns} "${newArtifactId}"`;
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: same

File: lib/modules/manager/sbt/update.ts (hunk 5 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/update.ts]: Good catch! Done :)

File: lib/modules/manager/sbt/update.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/modules/manager/sbt/update.ts:93]: My bad. We do use `escapeRegExp`

**Comment by gaeljw** [comment]
Comments resolved @viceice @RahulGautamSingh  :)

**Review by RahulGautamSingh — COMMENTED** [review]
In your test repo is the scalaVersion itself not supposed to get updated?

https://github.com/gaeljw/renovate-sbt-replacement-3/blob/3402cd0d381379a06ad8f6a6a2403e828a09fc89/build.sbt#L2

Or is that out of scope of this PR?

File: lib/modules/manager/sbt/update.ts [hunk 9]
```diff
@@ -32,3 +35,74 @@ export function bumpPackageVersion(
...
+}: UpdateDependencyConfig): string | null {
+  const { depName, newName, currentValue, newValue, sharedVariableName } =
+    upgrade;
+
+  let updatedContent = fileContent;
+
+  const [groupId, artifactId] = depName!.split(':', 2);
+
+  // Update the version
+  if (newValue && currentValue && newValue !== currentValue) {
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts:51]: this can also be true for a replacement, so you should check that first

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/sbt/update.ts (hunk 9 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/update.ts:51]: I'm not sure what you mean. It's intended that this `if` applies also for a replacement. I've added tests for the various cases, like `"should do replacement with new value`" and `"should do replacement without a new value"`.

**Comment by gaeljw** [comment]
> In your test repo is the scalaVersion itself not supposed to get updated?
> 
> https://github.com/gaeljw/renovate-sbt-replacement-3/blob/3402cd0d381379a06ad8f6a6a2403e828a09fc89/build.sbt#L2
> 
> Or is that out of scope of this PR?

Very good catch! I think there was no possible upgrade when I started this PR but now there is and it's not handled correctly.

```
 WARN: Error updating branch (repository=gaeljw/renovate-sbt-replacement-3, branch=renovate/scala-2.13.x)
       "err": {
         "message": "Cannot read properties of undefined (reading 'replace')",
         "stack": "TypeError: Cannot read properties of undefined (reading 'replace')\n    at escapeRegExp (file:///home/gaeljw/renovate/lib/util/regex.ts:72:16)\n    at updateDependency (file:///home/gaeljw/renovate/lib/modules/manager/sbt/update.ts:76:57)\n    at getUpdatedPackageFiles (file:///home/gaeljw/renovate/lib/workers/repository/update/branch/get-updated.ts:261:30)\n    at processTicksAndRejections (node:internal/process/task_queues:105:5)\n    at processBranch (file:///home/gaeljw/renovate/lib/workers/repository/update/branch/index.ts:605:19)\n    at instrument.attributes (file:///home/gaeljw/renovate/lib/workers/repository/process/write.ts:187:21)\n    at writeUpdates (file:///home/gaeljw/renovate/lib/workers/repository/process/write.ts:160:17)\n    at update (file:///home/gaeljw/renovate/lib/workers/repository/process/extract-update.ts:253:11)\n    at Module.renovateRepository (file:///home/gaeljw/renovate/lib/workers/repository/index.ts:143:19)\n    at instrument.attributes (file:///home/gaeljw/renovate/lib/workers/global/index.ts:219:11)\n    at start (file:///home/gaeljw/renovate/lib/workers/global/index.ts:204:7)\n    at file:///home/gaeljw/renovate/lib/renovate.ts:20:22"
       }
```

I missed that Scala versions upgrades are also handled through this piece of code. I'll a test to cover that and fix it.

File: lib/modules/manager/sbt/update.spec.ts [hunk 10]
```diff
@@ -48,4 +49,353 @@ describe('modules/manager/sbt/update', () => {
...
+          newValue: '1.0.1',
+        },
+      });
+
+      expect(res).toEqual(
+        codeBlock`addSbtPlugin("org.example" % "foo" % "1.0.1")`,
+      );
+    });
+
+    it('should update Scala version', () => {
```
Comment (gaeljw) [inline lib/modules/manager/sbt/update.spec.ts:130]: Well this is actually challenging and makes me wonder if the approach is right here because the `upgrade` content for a Scala version upgrade is quite different from a regular dependency:
```js
{
          depName: 'scala',
          packageName: 'org.scala-lang:scala-library',
          currentValue: '2.13.17',
          newValue: '2.13.18',
}
```
I cannot use neither the `depName` nor the `packageName` to find out where to do the modifications. I'm considering adding a special case when packageName = `org.scala-lang:scala-library` but this is not super satisfying.

Do you have any suggestion @RahulGautamSingh or @viceice ?

I've added this failing test in the meantime.

Comment (RahulGautamSingh) [reply lib/modules/manager/sbt/update.spec.ts:130]:
```
// Handle Scala version updates (depName is 'scala', not 'groupId:artifactId')
  if (depName === 'scala') {
    if (newValue && currentValue && newValue !== currentValue) {
      const pattern = regEx(
        `(scalaVersion\\s*:=\\s*)"${escapeRegExp(currentValue)}"`,
        'g',
      );
      updatedContent = updatedContent.replace(pattern, `$1"${newValue}"`);
    }
    return updatedContent;
  }
```

This works? ie. handling `scala` early.

Comment (gaeljw) [reply lib/modules/manager/sbt/update.spec.ts:130]: Oh right, handling it early does the job. I pushed a commit to do so.

I've also added handling Scala version defined through a shared variable (for that I had to update extraction as well)

File: lib/modules/manager/sbt/extract.ts [hunk 11]
```diff
@@ -82,6 +83,11 @@ const scalaVersionMatch = q
         separateMinorPatch: true,
       };
       ctx.scalaVersion = normalizeScalaVersion(ctx.scalaVersion);
+      if (ctx.variableName) {
```
Comment (gaeljw) [inline lib/modules/manager/sbt/extract.ts:112]: ℹ️ I could move that to a dedicated pre-requisite PR if you prefer.

**Comment by gaeljw** [comment]
I just came upon https://github.com/renovatebot/renovate/pull/40699 and I think my changes would make the latter not working anymore, it would need special handling in the `update.ts` to handle `sharedVariable` with a dot to find out where to do the upgrade. I didn't investigate what makes it work today with the generic upgrade method, maybe there are some things to copy/paste. cc @Thooms in case you have some feedbacks.

**Review by viceice — COMMENTED** [review]
looks good. please fix lint issues

**Comment by gaeljw** [comment]
> looks good. please fix lint issues

I don't see any lint issue locally or here in the CI. Am I missing something?

**Review by RahulGautamSingh — APPROVED** [review]

File: lib/modules/manager/sbt/extract.ts [hunk 12]
```diff
@@ -108,6 +109,11 @@ const scalaVersionMatch = q
...
+        dep.sharedVariableName = ctx.variableName;
+        dep.variableName = ctx.variableName;
+        delete ctx.variableName;
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.ts:115]: something looks wrong here

File: lib/modules/manager/sbt/update.ts [hunk 13]
```diff
@@ -32,3 +35,108 @@ export function bumpPackageVersion(
...
+          { depName, sharedVariableName, currentValue, newValue },
+          'Updating Scala version in shared variable',
+        );
+
+        const patternSharedVariable = regEx(
+          `${escapeRegExp(sharedVariableName)}(\\s*:\\s*String)?\\s*=\\s*"${currentValue}"`,
+          'g',
+        );
+
+        updatedContent = updatedContent.replace(
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: you're using global regEx  flag, should we use replace all here?

otherwise remove the global flag

File: lib/modules/manager/sbt/update.ts [hunk 14]
```diff
@@ -32,3 +35,108 @@ export function bumpPackageVersion(
...
+      } else {
+        logger.debug(
+          { depName, currentValue, newValue },
+          'Updating Scala version',
+        );
+        const pattern = regEx(
+          `(scalaVersion\\s*:=\\s*)"${escapeRegExp(currentValue)}"`,
+          'g',
+        );
+        updatedContent = updatedContent.replace(pattern, `$1"${newValue}"`);
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: same here

File: lib/modules/manager/sbt/update.ts [hunk 15]
```diff
@@ -32,3 +35,108 @@ export function bumpPackageVersion(
...
+        { depName, currentValue, newValue },
+        'Updating version for sbt dependency',
+      );
+
+      const pattern = regEx(
+        `"${escapeRegExp(groupId)}"\\s*(%|%%|%%%)\\s*"${escapeRegExp(artifactId)}"\\s*%\\s*"${escapeRegExp(currentValue)}"`,
+        'g',
+      );
+
+      updatedContent = updatedContent.replace(pattern, (match) => {
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts]: same here

File: lib/modules/manager/sbt/update.ts [hunk 16]
```diff
@@ -32,3 +35,108 @@ export function bumpPackageVersion(
...
+    logger.debug(
+      { depName, newName, currentValue, newValue },
+      'Applying replacement update for sbt dependency',
+    );
+
+    const [newGroupId, newArtifactId] = newName.split(':', 2);
+
+    const pattern = regEx(
+      `"${escapeRegExp(groupId)}"\\s*(%|%%|%%%)\\s*"${escapeRegExp(artifactId)}"`,
+      'g',
```
Comment (viceice) [inline lib/modules/manager/sbt/update.ts:131]: same

File: lib/modules/manager/sbt/extract.ts (hunk 12 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/extract.ts:115]: What do you mean? I applied similar logic to what is already done in some lines below:
- https://github.com/renovatebot/renovate/blob/c933fd6669bddff2bf34ff07f23642b88cbdf602/lib/modules/manager/sbt/extract.ts#L274-L279
- https://github.com/renovatebot/renovate/blob/c933fd6669bddff2bf34ff07f23642b88cbdf602/lib/modules/manager/sbt/extract.ts#L299-L302

Without the `delete`, the `variableName` (and other items of the context) "persists" across unrelated dependencies IIRC.

File: lib/modules/manager/sbt/update.ts (hunk 13 above)
Comment (gaeljw) [reply lib/modules/manager/sbt/update.ts]: Good catch, I switched to replaceAll and there was already a test covering the corner case of the same dependency appearing multiple times.
